python/ray/data/_internal/logical/optimizers.py

```diff
@@ -19,30 +30,211 @@ class PhysicalOptimizer(Optimizer):
+                    != down_logical_op._fn_constructor_kwargs
+                )
+            )
+        ):
+            return False
+
+        # Only fuse if the ops' remote arguments are compatible.
+        if not _are_remote_args_compatible(
+            up_logical_op._ray_remote_args or {}, down_logical_op._ray_remote_args or {}
+        ):
+            return False
+
+        # Otherwise, ops are compatible for fusion.
+        return True
+
+    def _fuse(self, down_op: PhysicalOperator, up_op: PhysicalOperator):
+        """Fuse the downstream operator with its upstream operator."""
+        from ray.data._internal.execution.operators.map_operator import MapOperator
+        from ray.data._internal.logical.operators.map_operator import AbstractMap
+
+        # Fuse operator names.
+        name = up_op.name + "->" + down_op.name
+
+        down_logical_op = self._op_map.pop(down_op)
+        up_logical_op = self._op_map.pop(up_op)
+
+        # Merge target block sizes.
+        down_target_block_size = down_logical_op._target_block_size
+        up_target_block_size = (
+            up_logical_op._target_block_size
+            if isinstance(up_logical_op, AbstractMap)
+            else None
+        )
+        if down_target_block_size is not None and up_target_block_size is not None:
+            target_block_size = max(down_target_block_size, up_target_block_size)
+        elif up_target_block_size is not None:
+            target_block_size = up_target_block_size
+        else:
+            target_block_size = down_target_block_size
+
+        # Fuse transformation functions.
+        down_transform_fn = down_op.get_transformation_fn()
+        up_transform_fn = up_op.get_transformation_fn()
+
+        def transform_fn(blocks: Iterator[Block]) -> Iterator[Block]:
+            blocks = up_transform_fn(blocks)
+            # TODO(Clark): Add zero-copy batching between transform functions.
+            return down_transform_fn(blocks)
+
+        # We take the downstream op's compute in case we're fusing upstream tasks with a
+        # downstream actor pool (e.g. read->map).
+        compute = get_compute(down_logical_op._compute)
+        ray_remote_args = down_logical_op._ray_remote_args
+        # Make the upstream operator's inputs the new, fused operator's inputs.
+        input_deps = up_op.input_dependencies
+        assert len(input_deps) == 1
+        input_op = input_deps[0]
+
+        # Fused physical map operator.
+        op = MapOperator.create(
+            transform_fn,
+            input_op,
+            name=name,
+            compute_strategy=compute,
+            min_rows_per_bundle=target_block_size,
+            ray_remote_args=ray_remote_args,
+        )
+
+        # Build a map logical operator to be used as a reference for further fusion.
+        # TODO(Clark): This is hacky, remove this once we push fusion to be purely based
+        # on a lower-level operator spec.
```

Comment:
Given that we need information both at the `LogicalOperator` and `PhysicalOperator` level to perform this fusion, not sure what else we can do here other than:
1. Push the remaining requisite information into the physical `MapOperator` (constructor args, lift `target_block_size` out of the bundler, etc.).
2. Push the block transform function into the logical `AbstractMap` operator and do the fusion at the logical operator level.

Each have their complications/cons, e.g. for (1) we'd need to clear the constructor args at `op.start()` in order to keep us from needlessly hanging on to object references, and for (2) we'd be muddying the logical layer with an execution-level concept (block transformations).